latextools_utils/analysis.py

```diff
@@ -0,0 +1,426 @@
+            NO_BEGIN_END_COMMANDS | ONLY_COMMANDS_WITH_ARGS
+
+        Returns:
+        A list of all commands, which are preprocessed with the flags
+        """
+        # convert the filter into a function
+        if type(how) is str:
+            def command_filter(c): return c.command == how
+        elif type(how) is list:
+            def command_filter(c): return c.command in how
+        elif callable(how):
+            def command_filter(c): return how(c)
+        else:
+            raise Exception("Unsupported filter type: " + str(type(how)))
+        com = self._commands(flags)
+        return _copy_entries(filter(command_filter, com))
+
+    def _add_command(self, command):
+        self._all_commands.append(command)
+
+    def _build_cache(self, flags):
+        com = self._all_commands
+        if flags & ONLY_PREAMBLE:
+            def is_not_begin_document(c):
+                return not (c.command == "begin" and c.args == "document")
+            com = itertools.takewhile(is_not_begin_document, com)
+        for cflag in sorted(_FLAG_FILTER.keys()):
+            if flags & cflag:
+                f = _FLAG_FILTER[cflag]
+                com = filter(f, com)
+        self._command_cache[flags] = list(com)
+
+    def _commands(self, flags):
+        if flags not in self._command_cache:
+            self._build_cache(flags)
+        return self._command_cache[flags]
+
+
+def get_analysis(tex_root):
+    """
+    Returns an analysis of the document using a cache
+
+    Use this method if you want a fast result and don't mind if there
+    are small changes between the analysis and the usage.
+    Don't use this method if you ware looking forward in using
+    the regions of the commands (it will most likely not yield proper result)
+    (and is currently not supported with st2)
+
+    Arguments:
+    tex_root -- the path to the tex root as a string
+                if you use the view instead, the tex root will be extracted
+                automatically
+
+    Returns:
+    An Analysis of the view, which contains all relevant information and
+    provides access methods to useful properties
+    """
+    view = tex_root  # store for the case, that the analysis have to be done
+    if type(tex_root) is str or not _ST3 and type(tex_root) is unicode:
+        tex_root = tex_root
+    else:
+        tex_root = get_tex_root(tex_root)
+
+    try:
+        result = cache.read(tex_root, "analysis")
+    except cache.CacheMiss:
+        result = analyze_document(view)
+        # TODO
+        # cannot pickle sublime.Region in st2 this is a hacky workaround
+        # to remove those regions if someone is interested in the regions
+        # this method is not recommended anyways
+        if not _ST3:
+            for c in result._all_commands:
+                c.region = None
+                for k in _RE_COMMAND.groupindex.keys():
+                    region_name = k + "_region"
+                    c.__dict__[region_name] = None
```

Comment:
Since `Regions` are really just glorified tuples, it should be fairly easy to provide python with the knowledge of how to pickle / unpickle them using the [`copy_reg`](https://docs.python.org/2.6/library/copy_reg.html) module. Basically we'd need this code, probably in the `except` block at the start of the file:

```python
import copy_reg

def pickle_region(region):
    return (sublime.Region, (region.a, region.b))

copy_reg.pickle(sublime.Region, pickle_region)
```

I've smoke-tested that on ST2 and seems to pickle and unpickle regions successfully.